KAFKA-18168: Adding checkpointing for GlobalKTable during restoration…

janchilling opened the pull request:
… and closing.

To address the issue of not creating a checkpoint file during the restoring and closing process, called the GlobalStateUpdateTask.flushState() method in GlobalStateUpdateTask.initialize() and GlobalStateUpdateTask.close() methods. This will flush the state and create a checkpoint file thereby, avoiding the need to completely restore the entire state.

As discussed with @mjsax in the Jira ticket.

janchilling committed:
KAFKA-18168: Adding checkpointing for GlobalKTable during restoration and closing.

To address the issue of not creating a checkpoint file during the restoring and closing process, called the GlobalStateUpdateTask.flushState() method in GlobalStateUpdateTask.initialize() and GlobalStateUpdateTask.close() methods. This will flush the state and create a checkpoint file thereby, avoiding the need to completely restore the entire state.

mjsax commented:
@aliehsaeedii for review.

janchilling committed:
KAFKA-18168: Adding test cases for GlobalKTable during restoration and closing.

Added test cases to check if checkpointing happens during restoring and closing. Furthermore, updated the following test cases shouldNotCheckpointIfNotReceivedEnoughRecords(), shouldNotCheckpointIfNotReceivedEnoughRecords(), shouldNotCheckpointWhenFlushIntervalHasNotLapsed() to accept checkpointing during initialization.

aliehsaeedii commented:
Thanks, @janchilling, for raising the issue and the PR.
As discussed/agreed in the ticket, the plan was flushing and check-pointing during a successful execution of `reprocessState` or `restoreState` method in `GlobalStateManagerImpl` class to cover the instant restart and scaling out cases. Could you please explain how `GlovalStateUpdateTask.initialize()` covers these two cases?

janchilling commented:
Hi @aliehsaeedii,

Thank you for your response. During both restarting and scaling out, the `GlobalStreamThread` invokes the `GlobalStateUpdateTask.initialize()` method using `GlobalStateMaintainer` to properly restore the global state.
Ref - https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java#L244

And later as part of this process, either the `reprocessState` or `restoreState` methods will be executed, depending on the scenario:

- `restoreState` is called when simply reloading state.
- `reprocessState` is used when the global state needs to be reprocessed.

Additionally, the reason this solution logic is implemented in the `GlobalStateUpdateTask` class is to basically reuse the existing `flushState` method .

Let me know if you have any questions or if further clarification is needed!

aliehsaeedii reviewed (approved):
Thanks. LGTM!

aliehsaeedii commented:
@mjsax Approved (IMO)!

janchilling commented:
Hi @mjsax @aliehsaeedii ,

Seems as if the PR has become stale. Is there anything more I should do?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java:
```diff
@@ -96,6 +96,7 @@ public Map<TopicPartition, Long> initialize() {
         }
         initTopology();
         processorContext.initialize();
+        this.flushState();
```
```suggestion
       flushState();
```

janchilling replied:
Updated as per the request from the latest commit.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateUpdateTask.java:
```diff
@@ -138,6 +139,7 @@ public void flushState() {
     }
 
     public void close(final boolean wipeStateStore) throws IOException {
+        this.flushState();
```
```suggestion
       flushState();
```

mjsax replied:
It seems to not make sense to flush the state if `wipeStateStore == true` ? Not really a problem, as we are closing anyway, but it still feels like it's off.

janchilling replied:
Ohh! Thanks for pointing that out. It does seem unnecessary to call `flushState()` when `wipeStateStore == true` since the state will be deleted anyway.
I've updated the code to only `flushState()` when `wipeStateStore == false` from the latest commit.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java:
```diff
@@ -333,4 +343,21 @@ public void shouldWipeGlobalStateDirectory() throws Exception {
+    @Test
+    public void shouldCheckpointDuringClose() throws Exception {
+        globalStateTask.initialize();
+        globalStateTask.close(false);
```
Seems we need to reset both flags after `initialize()` but before `close()`?

janchilling replied:
Yeah missed it, now updated!

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java:
```diff
@@ -252,6 +252,10 @@ public void shouldCheckpointOffsetsWhenStateIsFlushed() {
     @Test
     public void shouldNotCheckpointIfNotReceivedEnoughRecords() {
         globalStateTask.initialize();
+        // Reset after initialization since checkpointing should happen during initialization, KAFKA-18168
```
```suggestion
        // Reset after initialization since checkpointing should happen during initialization
```

mjsax replied:
Or maybe even remove the commend entirely?

janchilling replied:
Updated as per the request from the latest commit.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/GlobalStateTaskTest.java:
```diff
@@ -252,6 +252,10 @@ public void shouldCheckpointOffsetsWhenStateIsFlushed() {
         globalStateTask.initialize();
+        // Reset after initialization since checkpointing should happen during initialization, KAFKA-18168
+        stateMgr.checkpointWritten = false;
+        stateMgr.flushed = false;
```
Nit: it might be slightly better, to do this a little later in the test, ie, just before we call `globalStateTask.maybeCheckpoint()`? Test logic might be easier to reads for humans as (1) reset flags, (2) checkpoint, (3) verify flags.

```
        globalStateTask.update(record(topic1, 1, currentOffsetT1 + 9000L, "foo".getBytes(), "foo".getBytes()));
        time.sleep(flushInterval); // flush interval elapsed
        
        stateMgr.checkpointWritten = false;
        stateMgr.flushed = false;
        
        globalStateTask.maybeCheckpoint();
        
        assertEquals(offsets, stateMgr.changelogOffsets());
        assertFalse(stateMgr.flushed);
        assertFalse(stateMgr.checkpointWritten);
```

Similar in other tests.

janchilling replied:
I've updated the test cases to match the above flow.

mjsax reviewed (commented):
Thanks for the PR. (4.0 release keeps us very busy, so things are slow...)

Overall LGTM, a few nits (eg, code style: we only use `this.` if necessary), and I think one test has a bug.

janchilling committed:
KAFKA-18168: Updated changes as per the reviews.

1. Removed this to maintain code style
2. Now conditionally flushing the state only if the wipeStateStore is false to avoid unnecessary call to flushState() method if wipeStateStore is true.
3. Removed comments from the test cases and updated the test cases as per the review

mjsax commented:
Thanks for the PR. Merged to `trunk`.

mjsax commented:
Cherry-picked to `4.0` and `3.9` branches.
